parfive/downloader.py

```diff
@@ -372,28 +374,114 @@ def callback(token, future, main_pb):
+                    # setup asyncio queue and writer
+                    queue = asyncio.Queue()
+                    download_workers = []
+                    writer = self.loop.create_task(
+                        self._write_worker(queue, file_pb, filepath))
+
+                    if max_splits and resp.headers.get('Accept-Ranges', None) == "bytes":
+                        # XXX: int(?)
+                        content_length = int(resp.headers['Content-length'])
+                        split_length = content_length // max_splits
+                        ranges = [
+                            [start, start + split_length]
+                            for start in range(0, content_length, split_length)
+                        ]
+                        # let the last part download everything
+                        ranges[-1][1] = ''
+                        for _range in ranges:
+                            download_workers.append(
+                                self.loop.create_task(self._http_download_worker(
+                                    session, url, chunksize, _range, timeout, queue, **kwargs
+                                ))
+                            )
+                    else:
+                        download_workers.append(
+                            self.loop.create_task(self._http_download_worker(
+                                session, url, chunksize, None, timeout, queue, **kwargs
+                            ))
+                        )
 
-                            # Update the progressbar for file
-                            if file_pb is not None:
-                                file_pb.update(chunksize)
+                    await asyncio.gather(*download_workers)
+                    await queue.join()
+                    writer.cancel()
+                    return str(filepath)
 
         except Exception as e:
             raise FailedDownload(filepath_partial, url, e)
 
-    @staticmethod
-    async def _get_ftp(session=None, *, url, filepath_partial,
+    async def _write_worker(self, queue, file_pb, filepath):
+        """
+
+        queue: `asyncio.Queue`
+             Queue for chunks
+
+        file_pb : `tqdm.tqdm` or `False`
+            Should progress bars be displayed for each file downloaded.
+
+        filepath: `pathlib.Path`
+            Path to the which the file should be downloaded.
+        """
+        with open(filepath, 'wb') as f:
+            while True:
+                offset, chunk = await queue.get()
+                # breakpoint()
+
+                f.seek(offset)
+                f.write(chunk)
+                f.flush()
+
+                # Update the progressbar for file
+                if file_pb is not None:
+                    file_pb.update(len(chunk))
+
+                queue.task_done()
+
+    async def _http_download_worker(self, session, url, chunksize, http_range, timeout, queue, **kwargs):
+        """
+        Worker for http requests.
```

Comment:
Can you expand the description of what it is doing here, i.e. how it interacts with the queue etc.